packages/data/src/components/with-select/with-select-new.js

```diff
@@ -0,0 +1,28 @@
+/**
+ * WordPress dependencies
+ */
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { useCallback } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import useSelect from '../use-select';
+
+const withSelect = ( mapSelectToProps ) => createHigherOrderComponent(
+	( WrappedComponent ) => ( ownProps ) => {
+		const mapSelect = useCallback(
+			( select, registry ) => mapSelectToProps(
+				select,
+				ownProps,
+				registry
+			),
+			[ ownProps ]
```

Comment:
Note, I could not do `Object.values( ownProps )` here because of this error:

> Warning: The final argument passed to useCallback changed size between renders. The order and size of this array must remain constant.

I experimented with generating a consistent array of values from the prop keys in the same order over multiple renders (and filling missing values with undefined) but this caused the same error (because it looks like the number of props can _increase_ between re-renders as well).  This almost smells like a problem with implementation in some component in the editor tree but I'm uncertain what to do here.  On the surface just passing in the prop object seems to be sufficient for this case.

Reply:
`ownProps` changes on every render so this is like not memoizing it and running the callback every time.

I think we can just wrap the inner component in a `React.memo`.

Updated: https://gist.github.com/epiqueras/7eae39ba6b903286cf17a4907902a630#file-with-select-js

Reply:
> ownProps changes on every render so this is like not memoizing it and running the callback every time.

I couldn't implement `React.memo` because it resulted in invalid blocks on existing content.  I'm not sure of the internal workings of it but it's a no go based on results.  So I implemented [`pure`](https://github.com/WordPress/gutenberg/blob/master/packages/compose/src/pure/index.js) which is should work similarly.

I'm not seeing _much_ difference in behaviour between the two when I do a rough comparison using the react dev profiler tool:

![with-use-callback](https://www.evernote.com/l/AAN3qvHdDUZEEZpEl8ABalGM0gPg82JCfSoB/image.png)

![with-pure](https://www.evernote.com/l/AAPvRIplBiVNPpDmXDMyLpQI4SEBEbkvy_YB/image.png)

However pure does seem marginally better so probably good to use it.

Reply:
`memo` doesn't implement `shouldComponentUpdate` like `pure` does, it just adds a special flag for the renderer to handle.

Both have the same functionality, but internally React has different rules for when to ignore a `shouldComponentUpdate` or a `memo`.

It sounds like these blocks are relying on some indeterministic and undocumented behavior.

Reply:
The blocks throwing the invalidation issue were basic blocks included with GB (paragraph block for instance).

Reply:
Could there be a state/props mutation somewhere in there?